pvanalytics/quality/time.py

```diff
@@ -33,3 +36,130 @@ def spacing(times, freq):
+    #     then `x` is rounded up, otherwise `x` is rounded down.
+    #     If not specified rounding will go up from `to` // 2.
+    up_from = up_from or to // 2
+    quotient, remainder = divmod(abs(x), to)
+    remainder[remainder > up_from] = to
+    remainder[remainder <= up_from] = 0
+    return np.sign(x) * (quotient*to + remainder)
+
+
+def shifts_ruptures(daytime, clearsky_midday, period_min=2,
+                    shift_min=15, round_up_from=None,
+                    prediction_penalty=13):
+    """Identify time shifts using the ruptures library.
+
+    Compares the solar-transit time from `clearsky_midday` with the
+    mid-point between sunrise and sunset time derived from the `daytime`
+    mask. These times are the times of the first and last True value on
+    each day.
+
+    The Pelt changepoint detection method is applied to the difference
+    in midday times derived from the `daytime` series and the expected
+    midday times from `clearsky_midday`. For each period between change
+    points the mode of the difference is rounded to a multiple of
+    `shift_min` and returned as the time-shift for all values in that
+    period.
+
+    Parameters
+    ----------
+    daytime : Series
+        Boolean time-series with True for day and False for night.
+    clearsky_midday : Series
+        Time of midday in minutes for each day with no time shifts
+        (i.e.based on solar position with a fixed-offset time zone).
+    period_min : int, default 2
+        Minimum number of days between shifts. Must be less than or equal to
+        the number of days in `daytime`.
+
+        Increasing this parameter will make the result less sensitive to
+        transient shifts. For example if your intent is to find and correct
+        daylight savings time shifts passing `period_min=60` can give good
+        results while excluding shorter periods that appear shifted.
+    shift_min : int, default 15
+        Minimum shift amount in minutes. All shifts are rounded to a multiple
+        of `shift_min`. [minutes]
+    round_up_from : int, optional
+        The number of minutes greater than a multiple of `shift_min` for a
+        shift to be rounded up. If a shift is less than `round_up_from` then
+        it will be rounded towards 0. If not specified then the shift will
+        be rounded up from `shift_min // 2`. Using a larger value will
+        effectively make the shift detection more conservative as small
+        variations will tend to be rounded to zero.
+    prediction_penalty : int, default 13
+        Penalty used in assessing change points.
+        See :py:method:`ruptures.detection.Pelt.predict` for more information.
+
+    Returns
+    -------
+    Series
+        Time shift in minutes at each timestamp in `daytime`.
+
+    Raises
+    ------
+    ValueError
+        If the number of days in `daytime` is less than `period_min`.
+
+    Notes
+    -----
+    Derived from the PVFleets QA project.
+
```

Comment:
This docstring would really benefit from an example, or, a paragraph describing a possible use of this function.

Tossing some text out:

Timestamped data from monitored PV systems may not always be localized to a consistent timezone. In some cases, data is timestamped with local time that may or may not be adjusted for daylight savings time transitions.  This function helps detect issues of this sort, by detecting points where the difference in times at a consistent point (solar noon) changes significantly. If the data's timestamps have not been adjusted for daylight savings transitions, the time of day at solar noon will change by roughly 60 minutes in the days before and after the transition.

To use this changepoint detection method to determine if your data's timestamps involve daylight savings transitions, first reduce your PV system data (irradiance or power) to a daily time series, with each point being the observed midday time in minutes. For example, if sunrise and sunset are inferred from the PV system data, the midday time can be inferred as the average of each day's sunrise and sunset time of day. To establish the expected midday time, calculate solar transit time in time of day. This function detects shifts in the difference between the observed and expected midday times, and returns  **(here I'm unclear what is being returned)**

Reply:
The return value is a series of minutes that if added to the timestamps of the series will effectively put the entire series into the same time zone as the midday times in `clearsky_midday` (to be renamed `transit`).

Currently returns a series of integers. Perhaps it would be better to return a series of `Timedelta`.

Reply:
I guess this speaks to the function's anticipated use: to detect timezone issues, or to correct them? Returning a series of integers or `Timedelta` is a correction, one would need to inspect the values to know if there is an issue. Maybe return both?  A Series of bool True for timestamps that are out of synch with expected, and a Series of suggested corrections. I'm not yet convinced that the corrections are desirable, since they rely on the observed transit values.

Reply:
Returning both is probably the most useful thing at this point. The issue with returning a boolean series, however, is that if I pass `transit` in central time, for example, and `midday` in Mountain time (with daylight savings) then every value will be flagged as shifted since there is a systematic 1 hour shift (and two hours during daylight time).

Reply:
I thought that the timezone of `clearsky_midday` (or `transit`) was assumed to be known. I'm not sure why a user would pass data with a different timezone. Maybe we need to talk with Matt again

Reply:
It is known, and the situation above is not likely, but I thought it was important to consider, since it is possible given that the time zone of `midday` (previously `daytime`) is not known.